packages/sacado/src/Sacado_Fad_Kokkos_Specialization.hpp

```diff
@@ -0,0 +1,306 @@
+
+    size_t num_teams = (total_extent + team_size - 1) / team_size;
+
+    Kokkos::parallel_for(
+        "Sacado::view_copy Hierarchical",
+        Kokkos::TeamPolicy<exec_space>(num_teams, team_size, vector_size),
+        KOKKOS_LAMBDA(
+            const typename Kokkos::TeamPolicy<exec_space>::member_type &team) {
+          size_t ii = team.league_rank() * team.team_size() + team.team_rank();
+          if (ii >= total_extent)
+            return;
+          // work around capture restriction
+          if (dst.data() == src.data())
+            return;
+          if constexpr (view_t::rank() == 0)
+            dst() == src();
+          else if constexpr (view_t::rank() == 1) {
+            dst(ii) == src(ii);
+          } else if constexpr (view_t::rank() == 2) {
+            int i1 = ii % src.extent(1);
+            int i0 = ii / src.extent(1);
+            dst(i0, i1) = src(i0, i1);
+          } else if constexpr (view_t::rank() == 3) {
+            int i2 = ii % src.extent(2);
+            int i1 = (ii / src.extent(2)) % src.extent(1);
+            int i0 = ii / (src.extent(2) * src.extent(1));
+            dst(i0, i1, i2) = src(i0, i1, i2);
+          } else if constexpr (view_t::rank() == 4) {
+            int i3 = ii % src.extent(3);
+            int i2 = (ii / src.extent(3)) % src.extent(2);
+            int i1 = (ii / (src.extent(3) * src.extent(2))) % src.extent(1);
+            int i0 = (ii / (src.extent(3) * src.extent(2) * src.extent(1)));
+            dst(i0, i1, i2, i3) = src(i0, i1, i2, i3);
+          } else if constexpr (view_t::rank() == 5) {
+            int i4 = ii % src.extent(4);
+            int i3 = (ii / src.extent(4)) % src.extent(3);
+            int i2 = (ii / (src.extent(4) * src.extent(3))) % src.extent(2);
+            int i1 = (ii / (src.extent(4) * src.extent(3) * src.extent(2))) %
+                     src.extent(1);
+            int i0 = (ii / (src.extent(4) * src.extent(3) * src.extent(2) *
+                            src.extent(1)));
+            dst(i0, i1, i2, i3, i4) = src(i0, i1, i2, i3, i4);
+          } else if constexpr (view_t::rank() == 5) {
+            int i5 = ii % src.extent(5);
+            int i4 = (ii / src.extent(5)) % src.extent(4);
+            int i3 = (ii / (src.extent(5) * src.extent(4))) % src.extent(3);
+            int i2 = (ii / (src.extent(5) * src.extent(4) * src.extent(3))) %
+                     src.extent(2);
+            int i1 = (ii / (src.extent(5) * src.extent(4) * src.extent(3) *
+                            src.extent(2))) %
+                     src.extent(1);
+            int i0 = (ii / (src.extent(5) * src.extent(4) * src.extent(3) *
+                            src.extent(2) * src.extent(1)));
+            dst(i0, i1, i2, i3, i4, i5) = src(i0, i1, i2, i3, i4, i5);
+          }
+        });
+    Kokkos::fence();
+    src = dst;
+  }
+}
+} // namespace Kokkos
+#endif // SACADO_VIEW_CUDA_HIERARCHICAL
```

Comment:
I know we discussed this before, but what what is the long-term plan for code in this file? I assume you want this kind of code to go away eventually?

Reply:
Yes. This is stuff we should look at how to get rid off. I have some ideas of how to proceed but I think we can do this after the Kokkos 5.0 release.